src/rcc.rs

```diff
@@ -164,231 +166,117 @@ impl HSEClock {
-
-                    if error < 0.0
-                        || plln * vco_clkin_mhz < 100.0
-                        || plln * vco_clkin_mhz > 432.0
-                        || plln < 50.0
-                        || plln > 432.0
-                    {
-                        core::u32::MAX
-                    } else {
-                        (error * 1_000.0) as u32
-                    }
-                })
-                .unwrap();
-
-            // PLLN corresponding to the best pllp_val
-            plln = ((sysclk_mhz * (pllp_val as f32)) / vco_clkin_mhz).floor();
-
-            // Pllp bits to be written in the register
-            let pllp = match pllp_val {
-                2 => 0b00,
-                4 => 0b01,
-                6 => 0b10,
-                8 => 0b11,
-                _ => unreachable!(),
-            };
-
-            // Update the real sysclk value
-            sysclk_mhz = (vco_clkin_mhz * plln) / (pllp_val as f32);
-            sysclk = (sysclk_mhz * 1_000_000.0) as u32;
-
-            // Turn PLL off
-            rcc.cr.modify(|_, w| w.pllon().off());
-            // Wait till PLL is disabled
-            while !rcc.cr.read().pllrdy().is_not_ready() {}
-
-            if self.hse.is_some() {
-                // If HSE is provided
-                // Configure PLL from HSE
-                rcc.pllcfgr.write(|w| unsafe {
-                    w.pllsrc()
-                        .hse()
-                        .pllm()
-                        .bits(pllm as u8)
-                        .plln()
-                        .bits(plln as u16)
-                        .pllp()
-                        .bits(pllp)
-                        .pllq()
-                        .bits(9)
-                });
-            } else {
-                // If HSE is not provided
-                // configure PLL from HSI
-                rcc.pllcfgr.modify(|_, w| unsafe {
-                    w.pllsrc()
-                        .hsi()
-                        .pllm()
-                        .bits(pllm as u8)
-                        .plln()
-                        .bits(plln as u16)
-                        .pllp()
-                        .bits(pllp)
-                });
-            }
+        let mut pll48clk_valid = false;
 
-            // Enable PLL
-            rcc.cr.modify(|_, w| w.pllon().on());
-            // // Wait for PLL to stabilise
-            while rcc.cr.read().pllrdy().is_not_ready() {}
+        if self.use_pll {
+            sysclk = base_clk / self.pllm as u32 * self.plln as u32
+                / match self.pllp {
+                    PLLP::Pllp2 => 2,
+                    PLLP::Pllp4 => 4,
+                    PLLP::Pllp6 => 6,
+                    PLLP::Pllp8 => 8,
+                };
 
-            use_pll = true;
+            if self.use_pll48clk {
+                pll48clk_valid =
+                    base_clk / self.pllm as u32 * self.plln as u32 / self.pllq as u32 == 48_000_000;
```

Comment:
According to the f4 hal code, USB specification allows +-0.25%

Reply:
Done.